### ISSUE TITLE
Multithreading

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#37]
I started some experimental multithreading code there:

https://github.com/thomaswaldmann/borg/tree/multithreading

especially:

https://github.com/ThomasWaldmann/borg/commit/240a27a22774eb055c7bff5097b4d6557bbb81c5

From Python c-api docs:
"the standard zlib and hashlib modules release the GIL when compressing or hashing data."
So current compression and hashing should be ok for good multithreading, as well as python's I/O (it releases GIL before I/O ops).

Additionally, later changesets implement code to release the GIL there:
- chunker
- crypto
- lz4 compression

---

:moneybag: [there is a bounty for this](https://www.bountysource.com/issues/17668505-multithreading-gil)

**Comment by ThomasWaldmann** [comment]
https://github.com/darkdragn/lz4tools/issues/4 see there for lz4.

**Comment by ThomasWaldmann** [comment]
2 things are TODO there:

- remote repo tests are broken (and thus currently disabled)
- that delayer thread is strange, better solution?

**Comment by enkore** [comment]
Just an idea

Traditional multi-threading (shared data, locking, queuing etc.) is often neither simple to develop nor test. I have been using ZeroMQ to effectively multi-thread things in a few situations now, and feel that it makes things much easier with essentially no relevant overhead (for inproc://), especially when using zerocopy.

I understand that in borg files are essentially processed by a *pipeline* of algorithms:

reading => chunking => hashing / encrypting => storage

This might lend itself well to the pipeline pattern in zmq, e.g. every one of these stages is handled by one or more nodes (=threads). Work distribution, queuing etc. are handled by zmq transparently.

This zmq approach to multithreading is not unlike microservices. It would probably mean extensive refactoring for Borg.

**Comment by ThomasWaldmann** [comment]
Interesting idea. Can you say what the pros of this approach are compared to Python's Queue module (which is in stdlib and threadsafe)? The Queue module is what the multithreading code uses currently.

**Comment by enkore** [comment]
Ah I see how you approached it, it is conceptually already very close. I also see the "delayer problem" now.

On a technical level the zmq approach is language agnostic, so it's [it can be] straightforward to implement entire stages in C with no Python/GIL involvement *at all* (and the ability to still do this zerocopy: separate, pure C-thread runs a node in the same process, so inproc:// can be used between Python and C parts, without ever copying the data).

**Comment by schuft69** [comment]
Regarding compression: 
We are using https://github.com/madler/pigz for parallel compressing of petabyte of archives. Maybe it could help here. We never had an issue  - and the performance on multicore systems is awsome!

**Comment by ThomasWaldmann** [comment]
@schuft69 please see the ticket about compression about why this is not that simple in case of borg.

**Comment by enkore** [comment]
Each chunk (0 byte ... 8 MB) is compressed independently, so parallel compressor tend to not work well there; instead we'll run separate stages (hashing, encryption, compression) in parallel and maybe multiple operations in parallel as well (e.g. compressing multiple chunks independently in parallel).

**Comment by sjuxax** [comment]
The multithreaded branch hasn't been updated since March 2016, about 1.25 years, and is missing about half of the commits in the current HEAD. How much of that approach remains viable?

Lack of some MT support makes borgbackup basically unusable for me. Even basic support would make things much more bearable (for example, compressing multiple chunks simultaneously).

I'd like to make contributions to help, but I'm not sure where to start right now. What's the current status on MT draft/direction?

**Comment by ThomasWaldmann** [comment]
@sjuxax that was just an experiment and should not be used. real, better working MT is planned for borg 1.2.

**Comment by enkore** [comment]
The multithreading branch was never intended to be merged; it was one of now several tests. The [Borg 1.2](https://github.com/borgbackup/borg/wiki/Borg-1.2) entry in the (project management) wiki might be of interest. Some further tests and planning has been conducted, but partly not published yet.

**Comment by sjuxax** [comment]
Thank you for the update. Is there a draft or issue discussing the current strategy for implementing MT support? I want to contribute something rudimentary, at least.

**Comment by ThomasWaldmann** [comment]
@sjuxax btw, why is it *unusable* without multithreading for you?

**Comment by sjuxax** [comment]
I'm mainly interested in using it to back up large, frequently-changing targets. Even if the job would normally complete without wrapping around on itself (i.e., the prior job wouldn't finish before the next interval), it's not worth the extra cost and risk to keep the backup going in the background for much longer than it would take other backup targets that do support multithreading to complete the job.

If I was backing up around 10G of stuff, it probably wouldn't be a big deal. I'm interested in backing up things that are much larger, almost always >= 100G with many small files. I have not tested lately, but back in April when I last tested, I definitely felt it was too slow for regular use.

As an example, I'm backing up a 5T filesystem right now (from disks that like to choke every few hours). While I recognize other limitations in borgbackup may make a 5T backup implausible, it would be convenient to be able to use it for more common tasks in the 100G-500G range, but without multithreading, it's just too slow for me.

I've tried several things over the last little bit, including SquashFS, fsarchiver, and others. Right now I am using ZFS on loopback with compression and deduplication enabled and rsyncing the tree over. That seems to be working the best, and it makes it easy to resume the process when the disks hang.

**Comment by enkore** [comment]
5T works quite well; the largest backup set I know of that was "publicly admitted" to use Borg is >40T. What's clear of course is that Borg 1.0 and 1.1 have fundamental limitations on their processing speed... whether this is a nasty problem for a big backup set has more to do with it's workload and less with the sheer size of it. Fast-changing sets require the most processing, while other workloads are usually less problematic. Files that don't change are faster to process than files that do; bigger files are slower than smaller files (<=one chunk) etc.

**Comment by ThomasWaldmann** [comment]
Well, you need to consider the amount of change between backups. If it is too slow to backup the changes, yeah, then you have a problem.

The total amount (as long as it is within the limitations) doesn't matter much, borg skips unchanged files rather quickly (be careful whether inode numbers are stable for your fs and always mount at same mountpoint).

Assuming you have enough backup space (after considering how much you safe due to historical dedup), you can use lz4 compression (you won't need multiple threads then to get compression fast). 

Encryption is also fast, IF you have AES-NI.

It's not perfect yet (thus our plans for 1.2), but quite ok - not only for small backups.

**Comment by ThomasWaldmann** [comment]
note: I'll start refactoring (slowly) for zeromq, based on current master code.

**Comment by enkore** [comment]
[Here's my prototype from March](https://github.com/enkore/borg/commit/4664f2df8bdf9a8d54e9fc18a571ca4e8a094957) which roughly corresponds to the current plan.

**Comment by ThomasWaldmann** [comment]
hmm, guess it makes more sense if you do a PR with that and I review it. And you could finish reviewing the crypto-aead stuff, so we can merge it first, so I don't need to rebase it again / fix conflicts again.

**Comment by ThomasWaldmann** [comment]
btw, the branch has only few changesets with huge changes. i imagined a bit smaller steps / less risk approach.

**Comment by wzyboy** [comment]
Regarding large (in size and/or in number of files) backups. Here is my use case.

In the past, I only used BorgBackup for not-very-busy servers. But recently I decided to try to back up a CI server. With lots of `node_modules/` (Node.js) and `vendor/` (Go) directories, the server has tons of small files.

The initial backup (~42 GiB) took only 42 minutes. That's about 1 GiB per minute. Considering the large number of files and network transfer time (I'm backing up from a data center near San Jose, CA to AWS us-east-1 N. Virginia region over SSH), it's quite impressive.

Here is the most recent backup:

```
Time (start): Tue, 2018-03-20 01:46:24
Time (end):   Tue, 2018-03-20 02:06:34
Duration: 20 minutes 9.91 seconds
Number of files: 6134998
Utilization of max. archive size: 2%
```

Even with single-core, the speed is acceptable. However, the fact that only one of the many cores of the CI server reaches 100% usage during the backup makes think "what a waste of CPU cores" :-)

**Comment by knutov** [comment]
What is the current state of this issue?

With `borg create --compression auto,zstd,3` on modern hardware (ssd) backup process speed is limited by speed of one cpu core. And it can be much faster on multiple cores.

**Comment by ThomasWaldmann** [comment]
See the milestones. Next release (1.2) will not yet have it.

It's quite a lot of work to restructure borg for this, better funding would be desirable also.

**Comment by alexander-rieder** [comment]
Any news on this?

**Comment by ThomasWaldmann** [comment]
No.

**Comment by tb25** [comment]
In shellpattern.py what do you mean by the TODO: support {alt1,alt2} shell-style alternatives. Could you please elaborate on this? What shell styles do you want to support? @ThomasWaldmann

**Comment by ThomasWaldmann** [comment]
@tb25 your question is completely unrelated to multithreading, so please move it to some more appropriate place (maybe a new ticket), so it can be deleted from here.

**Comment by br-olf** [comment]
Do you need help on this issue?

**Comment by ThomasWaldmann** [comment]
work on this (besides planning, experiments) has not started yet and likely some crypto stuff needs to be solved first (e.g. avoiding AES counter mode issues).

**Comment by br-olf** [comment]
OK sure.

Anyways here is my idea:
The chunker should fill multiple chunk queues, one for each thread we want to spawn.
Then we can process each queue with a separate thread.
This way we can parallelize the whole pipeline (hashing, cache lookup, compressing and maybe even encrypting).

This would require the cache to be thread save or read only and will probably result in RAM demands scaling with the number of threads.

**Comment by jean-phillipe88** [comment]
There is quite a nice bounty on this! I'd like to cash in on it. anyone thats been active for awhile can summarize what needs to be done? I have some quarantine time on my hands i would like to monetize on :P

**Comment by Justinzobel** [comment]
Added some more cash to the bounty. Hope we can get this feature soon!

**Comment by jean-phillipe88** [comment]
I'm ready to start work on it. I've sent Borg an email and waiting for a confirmation, but have yet to receive that.

**Comment by FelixSchwarz** [comment]
I'm TW (obviously) but my impression was that this issue ("multithreading") is mostly an epic "umbrella" issue. There are a lot of nuances involved (and probably some larger changes to borg's internal architecture).

I'm pretty sure the current bounty (USD 1500) is very very low compared to the regular rates of a professional software developer who needs to spend time on this (otherwise @ThomasWaldmann and others would have solved this a long time ago).

From what I can see #929 is a better place to start. The first step would be to define realistic intermediate goals (there are several operations which could benefit from using multiple CPU cores) and agree on ways to measure the impact.

**Comment by Justinzobel** [comment]
I say if Jean wants to give it a go and has nothing better to do then give it their best. :)

**Comment by jean-phillipe88** [comment]
Yeah, but i am an actual engineer. Would deff solve for the bounty, even if it takes me a whole week. So i mailed asking for detailed info on completion conditions etc to solve it properly. I will start once i have a reply and a go ahead on this platform :)

**Comment by jean-phillipe88** [comment]
> I'm TW (obviously) but my impression was that this issue ("multithreading") is mostly an epic "umbrella" issue. There are a lot of nuances involved (and probably some larger changes to borg's internal architecture).
> 
> I'm pretty sure the current bounty (USD 1500) is very very low compared to the regular rates of a professional software developer who needs to spend time on this (otherwise @ThomasWaldmann and others would have solved this a long time ago).
> 
> From what I can see #929 is a better place to start. The first step would be to define realistic intermediate goals (there are several operations which could benefit from using multiple CPU cores) and agree on ways to measure the impact.

Yeah that was my first impression as well. Would be easier to have clear and concise goals to claim to the bounty. While i am applying for jobs and interviews i could work on it. Seems right up my alley to be honest.

**Comment by FelixSchwarz** [comment]
> Would deff solve for the bounty, even if it takes me a whole week.

I won't claim much knowledge about borg's internals (I did contribute only tiny bits to borg) so maybe I'm wrong but I am pretty confident that a week is not enough. I guess with all the required architectural discussions, getting agreement on the preferred approach, coding, testing and benchmarking this is easily a month of work (at least).  I'd happy to be proved wrong though :-)

That's why working on smaller steps towards "multithreading" would be more beneficial. IMHO the first approach is unlikely to fully implement the desired features but with multiple developers each contributing some pieces we might get a version of borgbackup which utilizes the power of multicore machines.

Anyway: Outsiders discussing this in this ticket is unlikely to get real progress. If @jean-phillipe88  wants to start working on this I propose you read the relevant tickets in this tracker and then come up with a high-level step-by-step plan of the changes (and probably some more in-depth description of the necessary changes of the first step).

**Comment by jean-phillipe88** [comment]
> > Would deff solve for the bounty, even if it takes me a whole week.
> 
> I won't claim much knowledge about borg's internals (I did contribute only tiny bits to borg) so maybe I'm wrong but I am pretty confident that a week is not enough. I guess with all the required architectural discussions, getting agreement on the preferred approach, coding, testing and benchmarking this is easily a month of work (at least). I'd happy to be proved wrong though :-)
> 
> That's why working on smaller steps towards "multithreading" would be more beneficial. IMHO the first approach is unlikely to fully implement the desired features but with multiple developers each contributing some pieces we might get a version of borgbackup which utilizes the power of multicore machines.
> 
> Anyway: Outsiders discussing this in this ticket is unlikely to get real progress. If @jean-phillipe88 wants to start working on this I propose you read the relevant tickets in this tracker and then come up with a high-level step-by-step plan of the changes (and probably some more in-depth description of the necessary changes of the first step).

Yeah i can see why you think a month would be the minimum time duration using this approach. And also why the bounty is way to low to attract any serious interest. I would not really discuss any architectures or getting agreements, but just push to a fork for review to see if the results are worth it.

I'm used to being hired to upgrade someone else's code going in with zero knowledge beforehand. Never worked with large groups trying to contribute, and i see allot of problems getting even small changes would take forever using this approach.

But don't misunderstand me, i came here because of the bounty. If it is unlikely that i could earn it, i probably won't delve into it.

**Comment by jean-phillipe88** [comment]
I mean, this issue has been open for 5 years now. Is it ever meant to be marked as SOLVED?

**Comment by enkore** [comment]
Implementing multithreading isn't actually that much work, my prototype which mostly worked fine from a few years ago should still be in my repo – just a few hundred lines of diff or so. But actually making it go faster, especially when processing small chunks, is way more involved (pure Python isn't going to cut it, especially these days). IIRC it was about 20-30 % faster while using about 100 % more CPU when processing large files, and 20 % or so slower when processing small files. That would have been on a quad-core Xenon.

The design discussed 2016-2017 should be solid and scalable to 8+ cores, but many threads will need to run their consume/produce-loops without the GIL, i.e. only in native code. It's basically just a lot of plumbing and testing work.

**Comment by jean-phillipe88** [comment]
that is what i thought indeed.

**Comment by ThomasWaldmann** [comment]
the plan to work on multithreading is for `helium` milestone, after doing crypto work.

current master branch is still `hydrogen` and not released yet, but in alpha.

**Comment by jean-phillipe88** [comment]
Does that mean the bounty is currently inactive?

**Comment by ThomasWaldmann** [comment]
There is no "inactive" state for bounties. Once they are set, they are active.

But that does not necessarily mean they can be done at any time (without investing a lot of time on also solving the prerequisites).

Also, there are better and worse defined / scoped bounties and for new contributors I would rather suggest working on some smaller scope bounty / project.

There are lots of open tickets and this one is definitely one of the harder to solve ones.

**Comment by jean-phillipe88** [comment]
That is why I sent you an email requesting a clear and concise definition of the prerequisites for fulfilling the bounty requirements. I am a pro developer, and i came across this because of the bounty, and i am fairly confident in my ability to solve for it. 

Extra work on the way to it is possible, and if the work has bounties on the way, great; but honestly, i am a bounty hunter :)

**Comment by jean-phillipe88** [comment]
Did you get my email btw? 

> There is no "inactive" state for bounties. Once they are set, they are active.
> 
> But that does not necessarily mean they can be done at any time (without investing a lot of time on also solving the prerequisites).
> 
> Also, there are better and worse defined / scoped bounties and for new contributors I would rather suggest working on some smaller scope bounty / project.
> 
> There are lots of open tickets and this one is definitely one of the harder to solve ones.

**Comment by ThomasWaldmann** [comment]
If you are here just because of the bounty, I guess you better search for more attractive bounties elsewhere.

**Comment by jean-phillipe88** [comment]
I reacted to this one because from all the ones i checked out, this one seemed the fastest to accomplish. Just curious though, this bounty has been active for 5 years, is it never meant to be solved?

> If you are here just because of the bounty, I guess you better search for more attractive bounties elsewhere.

**Comment by ThomasWaldmann** [comment]
To all **backers of the bounty on bountysource** for this issue, you need to **urgently get active** or **your backing may get lost to bountysource**.

See #5230 for details (I posted a copy of my email there, which you can slightly modify and reuse for the email you need to write to them).

**Comment by ThomasWaldmann** [comment]
About my previous comment: no need to get active any more, this was resolved, see #5230.

**Comment by ThomasWaldmann** [comment]
As a general comment:

This is no easy issue and thus not suited for new contributors who are unfamiliar with the codebase and the required changes that need to implemented **before** even starting to work on **this** issue.

**Comment by l29ah** [comment]
Ouch.
Now it seems like rewriting borg in haskell would be easier than implementing the multithreading. At least, it seems like a manageable task for $1500.

**Comment by ThomasWaldmann** [comment]
@l29ah guess you're way to optimistic, but of course it depends on how quick you are in haskell and what you expect per hour.

if it is 0 per hour, you'll have infinite time. :-)

**Comment by l29ah** [comment]
The documentation of internals of borgbackup is just too good, i have to give writing a parallel borg-compatible backup tool a try :)

**Comment by enkore** [comment]
Do one better - learn from the issues documented in those docs and design something that avoids these issues ;)

**Comment by bluet** [comment]
No updates for a while here, hope everything's going well.  
Now people have more CPU cores but still stuck on one core.

(just added some more to the bounty)

**Comment by duven87** [comment]
What is the status of the issue?

**Comment by ThomasWaldmann** [comment]
Releasing 1.2 is the next goal, 1 milestone ticket still open.
After that there will be fixing of whatever was not discovered in beta/rc.
Then, some crypto improvements and then multithreading, see milestones.

**Comment by abebeos** [comment]
Re Bountysource, please be aware of this:

https://github.com/abebeos/main/issues/2

**Comment by enkore** [comment]
Bountysource issues are discussed at #5900 - please don't bump tickets with unrelated information.

**Comment by abebeos** [comment]
> Bountysource issues are discussed at #5900 - please don't bump tickets with unrelated information.

Well, aren't you two (commenter and "liker") something!

Next time try a reply like this:

"Thank you for taking the time to inform us. We have similar problems with bountysource, discussed within #5900)"

**Comment by ThomasWaldmann** [comment]
@abebeos as you can read in the other ticket, we are aware of the troubles with bountysource and not happy with how they do business. we appreciate that you informed us, but still i will need to clean up this ticket now and remove all off-topic comments, including ours.

**Comment by abebeos** [comment]
All good, I'll delete mine anyways!

**Comment by abebeos** [comment]
Can someone please update https://github.com/borgbackup/borg/projects/5, possibly adding more documentation to the top-level issue? Info like a very high-level processing suggestion, in the form

* prepare subsystems
  * migrate X to use library Y
  * convert Y to use library Z
  * convert Z to multithreading
* Task B
* Task C

**Comment by ThomasWaldmann** [comment]
We're short after borg 1.2 "hydrogen" release, so current focus is on:
- fixes for 1.2.x
- work on next milestone "helium" (mostly crypto and some more cleanups, likely borg 1.3 when released)

That work will take quite a while, so planning on next milestone "lithium" that comes after that also has quite some time still.

I intentionally split the multithreading stuff off into a separate milestone because that likely will cause quite a lot of changes and also potential bugs and profits from some crypto improvements done before it. Not having too much stuff in one milestone also makes more frequent releases possible (see how long 1.2.0 took after 1.1.0).

**Comment by abebeos** [comment]
This is a 7 year old issue. Possibly its time to just file/collect the relevant issues within https://github.com/borgbackup/borg/projects/5, and document the expected work as requested (at least a skeleton bullet-list, should not take more than an hour).

This way any interested party can work-in much faster, and possibly come up with a more flexible implementation path/plan.

**Comment by FabioPedretti** [comment]
Interesting approach by restic to scale tasks based on whether an operation is CPU or IO-bound: https://github.com/restic/restic/pull/3611

**Comment by boris22x** [comment]
> 5T works quite well; the largest backup set I know of that was "publicly admitted" to use Borg is >40T. What's clear of course is that Borg 1.0 and 1.1 have fundamental limitations on their processing speed... whether this is a nasty problem for a big backup set has more to do with it's workload and less with the sheer size of it. Fast-changing sets require the most processing, while other workloads are usually less problematic. Files that don't change are faster to process than files that do; bigger files are slower than smaller files (<=one chunk) etc.
> 
> But yeah, I'm aware of these limitations and I'll / we'll try to address them with 1.2.

I would love to use borg for all my backups, but I have ~300TB of data to be backed up and with the current single-core implementation it is just not feasible (weeks? months? no idea).

**Comment by l29ah** [comment]
I have multithreaded (hopefully) borg-compatible backups working in my implementation (https://github.com/l29ah/hyborg), but i got too lazy to write integration tests to verify that it backs up all the corner cases correctly just as borg itself. You can try it if you feel like.

**Comment by flxai** [comment]
@l29ah @ThomasWaldmann Thanks. Is it possible to build on this to create a PR?

**Comment by boris22x** [comment]
> @boris22x I guess you don't want to put 300T into a single repo. And if you have multiple/many repos, you can run multiple/many borg in parallel. Initial backup will still be an effort, but future backups will be quicker.
Why not? The 300TB of data is highly compressible so IMHO it makes sense to put it into one repository. If there are some technical reasons it would be good to state somewhere in the documentation that borgbackup supports backups up to xTB of data. The other solution I am using to back up the 300TB does the job well. The current single core limitation limits us to use it on a bigger scale (~5PB of data).

**Comment by ThomasWaldmann** [comment]
There is a documented archive size limitation (which is quite high, but check it yourself).

What I meant are rather practical problems, like e.g. borg check taking rather long for huge repos, borg client and server needing a lot of memory, repo locking, etc. - multiple smaller repos are just easier to deal with.

**Comment by boris22x** [comment]
I understand the possible practical problems with single-core implementation, it is not feasible to back up such huge repositories. But having a multicore system (28 or even 64 cores), a lot of RAM (we have a minimum of 2TB here) fast storage I do not really see the practical problems.
https://borgbackup.readthedocs.io/en/latest/faq.html#usage-limitations - nothing here
https://borgbackup.readthedocs.io/en/latest/internals/data-structures.html#indexes-caches-memory-usage - memory usage is explained here
https://borgbackup.readthedocs.io/en/1.0.11/internals.html - if I take the medium size of an item entry is 2kB (~100MB size files or more ACLs/xattrs), the limit will be ~32 million files/directories per archive resulting in maximum archive size 3PB+ which is for an enterprise backup sufficient. And yes, we can have thousands of small repositories because of software limitations or we use as before the other software.

**Comment by abebeos** [comment]
@boris22x , the problem with multithreading support is in the project-management.

Single issue, 8 years, project-lead did not manage to split this into multiple parts.

This is far too much centered around one person.

8 years no multi-threading is bluntly a failure (open source or not)

**Comment by RonnyPfannschmidt** [comment]
It's rare to see such malevolent and miss informed comments

**Comment by abebeos** [comment]
* Fact: the problem with multithreading support is in the project-management
* Fact: 8 years, project-lead did not manage to split this into multiple parts
* Fact: This is far too much centered around one person (clarification: the project, the processes)
* Fact: 8 years no multi-threading is bluntly a failure

And another fact, especially for you, @RonnyPfannschmidt

* Fact: contributors do not have enough professional pride to interfere in a way that project flaws are fixed.

**Comment by infectormp** [comment]
@abebeos I suspect you think this is easy. So, you can fork the project, roll it back 8 years and start adding multithreading yourself. Also, please rollback all libraries and python itself to an 8-year-old state.

**Comment by boris22x** [comment]
We made a very simple decision. Instead of trying to deal with low Borg backup performance related to single-core implementation, we switched to another solution that is capable to utilize all CPU cores. I took the same decision for my personal backups. I have to agree with @abebeos, having in 2023 a single-core implementation is a project-management issue where the project management clearly does not understand user requirements in 2023. If the Borg backup project management thinks that a single-core implementation is sufficient in 2023, then Borg backup could be used mostly by home users to backup a few GBs of data but not by power users requiring backing up several TBs of data and definitely not by enterprise requiring backing up several hundreds of TBs of data on a daily basis. @RonnyPfannschmidt I have no idea who defines priorities for Borg backup, but I would see instead of i.e. recompress functionality as a priority multi-core implementation. This is the way how I feel as a user that left Borg backup due to the missing multi-core implementation. Good luck!

**Comment by infectormp** [comment]
@boris22x and which solution are you using now?

**Comment by boris22x** [comment]
> @abebeos I suspect you think this is easy. So, you can fork the project, roll it back 8 years and start adding multithreading yourself. Also, please rollback all libraries and python itself to an 8-year-old state.

I do not think anyone thinks that this is easy, but I can say for myself I feel that this has had no priority for 8 years. While we have CPUs with 128 cores, Borg backup can utilize only 1 core. If you do not see this as a priority, OK.

**Comment by infectormp** [comment]
@boris22x CPU with 128 cores is expensive. Have you supported borg in any way, apart from criticism?

**Comment by abebeos** [comment]
> I suspect you think this is easy.

It is a matter of priority, mentality, professional pride.

> So, you can fork the project, roll it back 8 years and start adding multithreading yourself. Also, please rollback all libraries and python itself to an 8-year-old state.

You're talking nonsense.

**Comment by abebeos** [comment]
> CPU with 128 cores is expensive. Have you supported borg in any way, apart from criticism?

Fascinating logic...

**Comment by boris22x** [comment]
> @boris22x CPU with 128 cores is expensive. Have you supported borg in any way, apart from criticism?

So if a user will tell you that you have a problem you will tell them to F*CK OFF as they have not paid you? Do you realize that on your main project site, there is no support button that would allow anyone to send you i.e. money via Patreon or any similar platform? Do you realize that if I go to "Pay services" (https://www.borgbackup.org/support/commercial.html) I still do not see a way to support the project? I am using Hetzner, I was using borgbase.com (until I stopped because of the missing multi-core implementation), whether you get anything from them I have no idea. I understand that you feel criticized, but not everyone is a developer that could work on the project on the DEV site. And if you guys do not have a Support us by paying money on your main website then why do you complain about missing support?

**Comment by infectormp** [comment]
@boris22x and which solution are you using now?

**Comment by abebeos** [comment]
> I have no idea who defines priorities for Borg backup

@ThomasWaldmann

**Comment by infectormp** [comment]
> Do you realize that on your main project site, there is no support button that would allow anyone to send you i.e. money via Patreon or any similar platform?

https://github.com/borgbackup/borg#helping-donations-and-bounties-becoming-a-patron

**Comment by boris22x** [comment]
> @boris22x and which solution are you using now?

To be constructive here - we switched to hashbackups that seems to be capable to utilize more than a single core.

**Comment by boris22x** [comment]
> > Do you realize that on your main project site, there is no support button that would allow anyone to send you i.e. money via Patreon or any similar platform?
> 
> https://github.com/borgbackup/borg#helping-donations-and-bounties-becoming-a-patron

https://www.borgbackup.org/ - This is your main project site you get by google search "borgbackup", github is the DEV site. If not, then sorry, again, there is bad communication towards the users.

**Comment by boris22x** [comment]
> > Do you realize that on your main project site, there is no support button that would allow anyone to send you i.e. money via Patreon or any similar platform?
> 
> https://github.com/borgbackup/borg#helping-donations-and-bounties-becoming-a-patron

and one more thing - you even do not have a link from your main website to https://www.borgbackup.org/support/fund.html, this is only on github.

**Comment by boris22x** [comment]
> > > Do you realize that on your main project site, there is no support button that would allow anyone to send you i.e. money via Patreon or any similar platform?
> > 
> > 
> > https://github.com/borgbackup/borg#helping-donations-and-bounties-becoming-a-patron
> 
> https://www.borgbackup.org/ - This is your main project site you get by google search "borgbackup", github is the DEV site. If not, then sorry, again, there is bad communication towards the users.
Sorry, I found the button there...

**Comment by boris22x** [comment]
@ThomasWaldmann - what would be the amount you need to get the multi-core support done?

**Comment by RonnyPfannschmidt** [comment]
I stand by my comment and it's abundantly clear that spending any more time on the perpetrators is waste of it.

**Comment by abebeos** [comment]
> what would be the $$$ amount you need to get the multi-core support done?

Till Waldmann response, a bit background info:

- [A bounty of around $1580 exists (20 backers)](https://www.bountysource.com/issues/17668505-multithreading-gil)
- the effort is of course much higher
- A development mentality must exist, mainly to always split the system/work down into manageable units
  - this is currently not the case
- my estimation is that it would take around $1K to create just a work-plan (what needs to be done, in what sequence)
  - this needs collaboration (one abstract analyst + the main developer + some contributors)
    - my experience in this project tells me: this is not feasible
    
The biggest problem with the borg project is that it's arranged around the needs of its main developer (Waldman). It is not arranged in a way to serve external contributors (e.g. casual ones, that want to contribute just a bit for week).

Ask yourself a question: why, after 8 years, is there no work-plan available, that lists at least generally the necessary tasks needed to achieve multithreading?

**Comment by abebeos** [comment]
> I have multithreaded (hopefully) borg-compatible backups working in my implementation
source: https://github.com/borgbackup/borg/issues/37#issuecomment-1410368548

@l29ah, impressive. I assume that it would take less time to validate your implementation than to tweak borg-backup towards multithreading.

@boris22x , hashbackup looks actually very nice (especially the `-p0` option to... switch off multithreading). Why do you insist to enhance borg-backup? What does it have that hasbackup hasn't?

**Comment by deathtrip** [comment]
> hashbackup looks actually very nice (especially the `-p0` option to... switch off multithreading). Why do you insist to enhance borg-backup? What does it have that hasbackup hasn't?

Hashbackup is proprietary software, with a bunch of marketing speak and no documentation on the inner workings of it on their website.

**Comment by boris22x** [comment]
> I stand by my comment and it's abundantly clear that spending any more time on the perpetrators is waste of it.

It is demotivating to support you in any way if you ignore basic functionality requests and you think in 2023 there is no need for multicore implementation. But as mentioned above I already moved on due to this limitation.

**Comment by boris22x** [comment]
> > I have multithreaded (hopefully) borg-compatible backups working in my implementation
> > source: [#37 (comment)](https://github.com/borgbackup/borg/issues/37#issuecomment-1410368548)
> 
> @l29ah, impressive. I assume that it would take less time to validate your implementation than to tweak borg-backup towards multithreading.
> 
> @boris22x , hashbackup looks actually very nice (especially the `-p0` option to... switch off multithreading). Why do you insist to enhance borg-backup? What does it have that hasbackup hasn't?

After the comment from Ronny I do not plan to spend any single $ supporting borgbackup. I wanted to do it because of nostalgy, but if someone is just blind and telling people to f*ck off when they tell them that the software does not meet 2023 requirements it is not worth a single cent.

**Comment by boris22x** [comment]
> > hashbackup looks actually very nice (especially the `-p0` option to... switch off multithreading). Why do you insist to enhance borg-backup? What does it have that hasbackup hasn't?
> 
> Hashbackup is proprietary software, with a bunch of marketing speak and no documentation on the inner workings of it on their website.

While borgbackup is stuck with the functionality in the year 2000, hashbackup is trying to solve the problems users have. If I cannot use borgbackup to backup my data at home or in the company I will look for another working solution. And I wonder why you keep ignoring your users, but it is your choice guys. Good luck with that approach and continue to use your single core CPUs and floppy drives... ;)  Unsubscribing from the discussion.

**Comment by abebeos** [comment]
> After the comment from Ronny

just a contributor, not the lead.

**Comment by abebeos** [comment]
> continue to use your single core CPUs and floppy drives... ;)

good one!!!

**Comment by l29ah** [comment]
> > I have multithreaded (hopefully) borg-compatible backups working in my implementation
> > source: [#37 (comment)](https://github.com/borgbackup/borg/issues/37#issuecomment-1410368548)
> 
> @l29ah, impressive. I assume that it would take less time to validate your implementation than to tweak borg-backup towards multithreading.

Try it. I think so as well, but i'm not a Python developer.

**Comment by lassepe** [comment]
@abebeos @boris22x. You are flooding the inbox of 21+ people that are subscribed to this issue. **Please: stay on topic and watch your language.** Both of your are being unnecessarily rude and aggressive in this thread.

Your behavior does not contribute to this issue being fixed any earlier; quite the opposite: with your behavior you make this thread unpleasant to read and even more unpleasant to work on. Open-source developers are humans and most people don't feel encouraged to work on something because of a random persons rant. Please, [don't be that open-source user](https://jacobtomlinson.dev/posts/2022/dont-be-that-open-source-user-dont-be-me/).

**Comment by abebeos** [comment]
@lassepe , what kind of clown are you? Get serious, and know who to patronize, you joker.

You FUCKING joker.

**Comment by abebeos** [comment]
> Try it. I think so as well, but i'm not a Python developer.

Nah, I was just here to work on some bounties, but borg-backup is a sinking ship.

Terribly amateurish processes and contributors.

Me following @boris22x, and out of here.

**Comment by abebeos** [comment]
(and spare your childish thumb-downs, I delete my comments anyways at some point)

**Comment by FelixSchwarz** [comment]
> Nah, I was just here to work on some bounties, but borg-backup is a sinking ship.
> 
> Terribly amateurish processes and contributors.
> 
> Me following @boris22x, and out of here.

Thank you @abebeos for making your point. Now I think we can go back to actually using this issue to track progress/share information about borgbackup's multithreading implementation.

**Comment by abebeos** [comment]
> Now I think we can go back to actually using this issue to track progress/share information about borgbackup's multithreading implementation.

what progress? You all completely delusional?

This issue is overdue to be closed & replaced by a follow-up issue (that mentions only relevant information).

**Comment by abebeos** [comment]
Ah, its fascist time. Well, well...

**Comment by seeker** [comment]
To any repository owners / moderators: Perhaps [hide](https://docs.github.com/en/communities/moderating-comments-and-conversations/managing-disruptive-comments#hiding-a-comment) off topic comments on this issue (including my own responses) to keep it clean?

**Comment by lazaridiscom** [comment]
anyone can hide messages themselves.

And such a gigantic, 8 years old issue is only kept open by amateurs.

Close it, and get a new one up, with fresh info, thus interested parties do not need to read through outdated junk.

**Comment by abebeos** [comment]
> the plan to work on multithreading is for `helium` milestone, after doing crypto work.
https://github.com/borgbackup/borg/issues/37#issuecomment-640053227

Milestones are a unclear mess:

https://github.com/borgbackup/borg/milestones


> There is no "inactive" state for bounties. Once they are set, they are active.
> 
> But that does not necessarily mean they can be done at any time (without investing a lot of time on also solving the prerequisites).
https://github.com/borgbackup/borg/issues/37#issuecomment-640106691

The main failure of the team-lead and contributors:

the "prerequisites" need to be listed.

(me now shredding all my comments)

**Comment by grinapo** [comment]
A few sidenotes (with preissued apology to all the readers of this issue).

- Issues (and possibly part of the issues) can be converted (or directed to) **discussions**. It may be much better way to actually _discuss_ the topics related to, but not directly affecting, this issue.
- *Hiding* discussion comments may be helping onlookers (like myself) to see how the actual issue is progressing.
- Could someone **summary where multithreading [or its plans] stand as of 2023**? I am right now looking at a `prune`, running on 100% cpu for more than 3 hours now, with about 5% disk utilisation. (Granted, info says that the repo contains about 5PB of data, which is both amusing and scaring, details are in https://github.com/borgbackup/borg/discussions/7766)

**Comment by enkore** [comment]
> summary where multithreading [or its plans] stand as of 2023?

Nowhere in particular. Note that most discussions and plans, and also the prototypes, are for `borg create` only, not other operations like `borg prune` (which can also be parallelized to some extent, but has non-trivial time/space trade-offs and other issues, primarily prefetching).

**Comment by ThomasWaldmann** [comment]
multithreading (and maybe other parallelism) is next milestone after borg2.

**Comment by struthio** [comment]
> multithreading (and maybe other parallelism) is next milestone after borg2.

Any time estimate ?

**Comment by ThomasWaldmann** [comment]
borg2 might take quite a while **if** some storage related changes get implemented before release (see "breaking" label). but these changes likely would be also useful for more parallelism.

**Comment by debuglevel** [comment]
just a side note for the few guys using borg in Windows:
some kind of parallelism might also greatly improve `borg` on Windows. I did not do any comparisons and benchmarks (so please be gentle :D), but NTFS seems to be painfully slow on single-threaded tools.

I'm also using Microsoft `Robocopy` (which seems to be a rather weird `rsync` clone) to synchronize around a million files once an hour - and it is actually quite fast (38 minutes). Turns out, it spins up 8 threads by default ;-).

**Comment by ThomasWaldmann** [comment]
Note: see #8217 benchmark results about why compressor-builtin parallelism does not work for us.

**Comment by struthio** [comment]
> Note: see #8217 benchmark results about why compressor-builtin parallelism does not work for us.

I guess this used just multi-threading in compression algorithm. But where might be other way to implement it which will improve performance.

Considering 10 Core CPU: When we backup folder that have like 10 files, then each file can be assigned to its own thread. Each file are processed in single core only, but whole folder goes in 10 threads in parallel. This should give a lot of performance increase.

**Comment by infectormp** [comment]
Just for info https://codspeed.io/blog/state-of-python-3-13-performance-free-threading

**Comment by SpiritInAShell** [comment]
> ... This should give a lot of performance increase.

(borg1.4.0) It seems that I can get a small performance boost an a single system _by accessing the local repository_ via `ssh://user@localhost`. IIRC! Therefor I will have 2 instances of borg running, one reading the data, the other handling repo access. RAM is no concern.

Any improvement on adding paralization would be great!

_While the CPU that runs the borg process maxes to 100%, the remaining 3 physical cores are not idling (as `btrfs` and `luks` are somewhat expensive) but not getting much to do, either._

_(I would only gain +10% single core performance, and +25% multi core (4:8 hyper threading) when upgrading to a E3-1280v2 from a E3-1225v2, which isn't financially feasable for that old machine.)_

**Comment by enkore** [comment]
In terms of hardware, getting a more recent CPU would make much more sense, especially since Borg benefits substantially from the SHA-NI extensions.

**Comment by ThomasWaldmann** [comment]
Update: there is some progress with "limited" multithreading (see PR links above):
- zstd has internal multithreading
- blake3 has internal multithreading
- we can easily overlap storing a pack and assembling the next pack - quite easy, no risk.

This will give us some speedups already and is way less complex / risky than a fully-queued fully-multithreaded borg code.

**Comment by ThomasWaldmann** [comment]
Closing this in favour of #10161, which collects everything from this thread that is still valid for borg2 (the pipeline design, enkore's prototype and its measured results, the GIL/native-code conclusion, the compressor-internal-MT findings, see also #8217 / #9961) plus what has meanwhile been implemented in master and the hard constraints that remain. This thread stays readable for the history.
